### PR TITLE
Add implementation for public order validation functions

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Adds 4 new public functions
    * `zeroEx.exchange.validateFillOrderAndThrowIfInvalidAsync`
    * `zeroEx.exchange.validateFillOrKillOrderAndThrowIfInvalidAsync` (calls the previous one under the hood)
    * `zeroEx.exchange.validateCancelOrderAndThrowIfInvalidAsync`
    * `zeroEx.exchange.isRoundingErrorAsync`

### CODE REVIEW

File: src/utils/order_validation_utils.ts [hunk 1]
```diff
@@ -1,10 +1,70 @@
...
+        if (wouldRoundingErrorOccur) {
+            throw new Error(ExchangeContractErrs.OrderFillRoundingError);
+        }
+    }
+    public async validateFillOrKillOrderThrowIfInvalidAsync(signedOrder: SignedOrder,
+                                                            fillTakerTokenAmount: BigNumber.BigNumber,
+                                                            takerAddress: string,
+                                                            zrxTokenAddress: string): Promise<void> {
+        await this.validateFillOrderThrowIfInvalidAsync(
+            signedOrder, fillTakerTokenAmount, takerAddress, zrxTokenAddress);
```
Comment (fabioberger) [inline src/utils/order_validation_utils.ts]: Put closing parens on new line.

File: src/utils/order_validation_utils.ts [hunk 2]
```diff
@@ -1,10 +1,70 @@
...
+        }
+    }
+    public async validateCancelOrderThrowIfInvalidAsync(order: Order,
+                                                        cancelTakerTokenAmount: BigNumber.BigNumber,
+                                                        unavailableTakerTokenAmount: BigNumber.BigNumber,
+    ): Promise<void> {
+        if (cancelTakerTokenAmount.eq(0)) {
+            throw new Error(ExchangeContractErrs.OrderCancelAmountZero);
+        }
+        const orderHash = utils.getOrderHashHex(order);
```
Comment (fabioberger) [inline src/utils/order_validation_utils.ts]: This variable is unused.

File: src/utils/order_validation_utils.ts [hunk 3]
```diff
@@ -1,10 +1,70 @@
...
+        if (order.takerTokenAmount.minus(unavailableTakerTokenAmount).eq(0)) {
```
Comment (fabioberger) [inline src/utils/order_validation_utils.ts]: Hm, can't we also just check for equality here?

File: test/order_validation_test.ts [hunk 4]
```diff
@@ -44,16 +46,112 @@ describe('OrderValidationUtils', () => {
...
     });
     beforeEach(async () => {
         await blockchainLifecycle.startAsync();
     });
     afterEach(async () => {
         await blockchainLifecycle.revertAsync();
     });
-    describe('#validateFillOrderBalancesAndAllowancesAndThrowIfInvalidAsync', () => {
+    describe('validateFillOrderAndThrowIfInvalidAsync', () => {
+        it('should throw when the fill amount is zero', async () => {
```
Comment (fabioberger) [inline test/order_validation_test.ts]: Should we also add test case that throws if the order has already been completely filled by someone else?

File: src/contract_wrappers/exchange_wrapper.ts [hunk 5]
```diff
@@ -161,7 +161,7 @@ export class ExchangeWrapper extends ContractWrapper {
         await assert.isSenderAddressAsync('takerAddress', takerAddress, this._web3Wrapper);
 
         const exchangeInstance = await this._getExchangeContractAsync();
-        await this._validateFillOrderAndThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, takerAddress);
+        await this.validateFillOrderThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, takerAddress);
```
Comment (LogvinovLeon) [inline src/contract_wrappers/exchange_wrapper.ts]: it's not just a wrapper. It passes the ZRX token address down.

File: src/utils/order_validation_utils.ts (hunk 1 above)
Comment (LogvinovLeon) [inline src/utils/order_validation_utils.ts]: Changed it here, but we have a ton of places where we lease the paren on the same line in tests.
It makes a huge difference, when assertions are 1-2 lines, not 1-3 lines.

**Pull request merged**